src/systemTest/java/com/hivemq/cli/utils/CLIShellTestExtension.java

```diff
@@ -0,0 +1,210 @@
+ * limitations under the License.
+ */
+
+package com.hivemq.cli.utils;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class CLIShellTestExtension implements BeforeEachCallback, AfterEachCallback {
+
+    public static final @NotNull List<String> CLI_EXEC =
+            Arrays.stream(Objects.requireNonNull(System.getProperty("cliExec")).split(" "))
+                    .collect(Collectors.toList());
+
+    private @Nullable Process cliShell;
+    private final @NotNull CommandConsumer commandConsumer = new CommandConsumer();
+    private final @NotNull CommandConsumer errorConsumer = new CommandConsumer();
+
+    @Override
+    public void beforeEach(final @NotNull ExtensionContext context) throws Exception {
+        final ArrayList<String> shellCommand = new ArrayList<>(CLI_EXEC);
+        shellCommand.add("sh");
+        cliShell = new ProcessBuilder(shellCommand).start();
+        waitForStartup(cliShell).get(3, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void afterEach(final @NotNull ExtensionContext context) {
+        if (cliShell == null) {
+            throw new IllegalStateException();
+        }
+        cliShell.destroy();
+    }
+
+    private @NotNull CompletableFuture<Void> waitForStartup(final @NotNull Process cliShell) {
+        final InputStream inputStream = cliShell.getInputStream();
+        final InputStreamReader inputStreamReader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
+        final BufferedReader bufferedInputReader = new BufferedReader(inputStreamReader);
+
+        final CompletableFuture<Void> cliReady = commandConsumer.waitFor("mqtt>");
+        final StringBuilder commandBuilder = new StringBuilder();
+
+        return CompletableFuture.runAsync(() -> {
+            while (!cliReady.isDone()) {
+                try {
+                    final int inputChar;
+                    inputChar = bufferedInputReader.read();
+
+                    if (inputChar == -1) {
+                        throw new IllegalStateException("End of stream was reached, but command was not found.");
+                    }
+                    commandBuilder.append((char) inputChar);
+                    commandConsumer.accept(commandBuilder.toString());
+                } catch (final IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }, runnable -> new Thread(runnable).start());
```

Comment:
Can be removed.

Reply:
This cannot be removed as the stream read action needs to be in a different thread. A thread pool would need to be cleaned after the tests which is not easy. So creating a single one time use thread is actually the cleanest way.